coalib/io/FileFactory.py

```diff
@@ -0,0 +1,115 @@
+import linecache
+import os
+
+from coala_utils.decorators import generate_eq
+
+
+@generate_eq('_filename', '_stamp')
+class FileFactory:
+    """
+    The FileFactory is a wrapper object to replace the actual file contents
+    from the file-dict. The object contains the following information about
+    the file:
+
+        * The filename (absolute file path).
+        * The parent directory the file is located in.
+        * The modification timestamp of the file.
+        * The file in the form of a string.
+        * The file in the form of a list.
+        * The file in binary format.
+
+    The equality of the object is checked just based on its filename and
+    timestamp.
+
+    To initialize a FileFactory object for a file:
+
+    >>> fp = FileFactory("./tests/io/FileFactoryTestFiles/test.txt")
+
+    Make the object into an iterable when required
+
+    >>> for line in fp:
+    ...     print(line)
+    This is a test file.
+    <BLANKLINE>
+
+    Retrieve a single line:
+
+    >>> fp.line(1)
+    'This is a test file.\\n'
+
+    Get all the lines in the file as a list:
+
+    >>> fp.lines
+    ('This is a test file.\\n',)
+    """
+
+    def __init__(self, filename):
+        """
+        :param filename:
+            The absolute filepath.
+        """
+        self._filename = os.path.abspath(filename)
+        self._parent = os.path.abspath(os.path.dirname(self._filename))
+        self._stamp = os.path.getmtime(self._filename)
```

Comment:
Why do we need this as data member ?
This might not be same when we want to fetch it. And we'll again have to call `getmtime` function.
I don't see any point in storing this.

Reply:
I see your point but @Makman2 also wanted to compare factory objects based on their filename and timestamp both

If we are going to compare the object based on the timestamp then we would need this as a data member.

Maybe I can move this to this data member to its own function that way whenever it is called it will store the most recent time stamp but I wonder whether moving this out of the `__init__` method would still make the `@generate_eq(_stamp)` work or not.

Reply:
I just tested moving the timestamp to its own function and it causes problems with `generate_eq('_filename', '_stamp')` because it cannot find `_stamp` in the `__init__` method

Reply:
The `timestamp` is essential to caching. It effectively encodes information if a file has been changed or not.

`timestamp` might not be up-to-date, but it will be at least the earliest state we can access. We can always have external programs modifying the file in the meanwhile, and by using the timestamp as early as possible, we have a safety measure that automatically causes our caches to revalidate.

Reply:
@palash25 regarding `generate_eq`: Just use the property names, I guess that works better